design-proposals/kubevirt_ephemeral_csi_driver.md

```diff
@@ -0,0 +1,79 @@
+
+## Motivation
+
+The ephemeral csi driver will allow one to specify [csi ephemeral volumes](https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#csi-ephemeral-volumes) inside a pod spec. Csi ephemeral volumes act very similar to an [emptyDir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) in that they are created when a pod starts, and will be removed when the pod stops. Like emptyDir, the csi ephemeral volumes provide a stable path for the volume inside a pod (like the virt-launcher), that is also visible on the node, the pod is running on. This allows the node (or a privileged pod on that node that has the host file system mounted, like the virt-handler pod) to manipulate the contents of the path and have them show up inside the pod. This is important for features like volume hotplug that bind mount volumes into the virt-launcher pod using that volume path. 
+
+Another property of csi ephemeral volumes is that they can be specified inline in a pod spec and do not require an explicit persistent volume to function. When a pod using the csi ephemeral volume terminates, the csi driver is free to clean up the contents of the ephemeral volume, or perform any other operation it wants on the ephemeral volume.
+
+As said above a csi ephemeral volume is very close in function to an emptyDir, so why not simply use an emptyDir? The main motivation in suggesting an ephemeral csi driver is control over what happens when a pod exits. The following scenario can happen when using an emptyDir in combination with volume hotplug:
+
+1. We have a running VM, and have hotplugged one or more volumes into it. We have used an emptyDir as a stable path inside the virt-launcher which we used to attach the disks in those volumes to the VM (this is what the current volume hotplug implementation does).
+2. Something causes the virt-launcher to be force deleted (user action, resource pressure, whatever)
+3. When we normally terminate a VM, the virt-handler is notified we are going to terminate the VM, and is given a chance to unmount any volumes before the virt-launcher pod exits.
+4. When a pod that uses an emptyDir terminates, the kubelet will remove all the contents of the volume backing the emptyDir.
+5. Because we force terminated the virt-launcher pod, we are now in a race between the kubelet, and the virt-handler (normally the grace period will give virt-handler enough time). If the virt-handler is called to unmount and hotplugged volumes before the kubelet clears the emptyDir, we are fine. However if the kubelet is run before the virt-handler, it will blindly remove the contents of the emptyDir volume, which, in the hotplug volume case will be the disk.img files that have not been unmounted. Since it is a bind mount this will result in the disk.img file on the actual volumes being removed, and we have lost data.
```

Comment:
Could you elaborate this a little bit more? From my experience in such cases the kubelet would fail to do the actual delete because a "device" (themounte disk) is busy.

Reply:
Just to clarify what I am talking about. Here an example where I mount a cirros disk to `test1234` and then try to delete it:

```bash
$ sudo mount -o rw,bind $PWD/cirros-0.5.2-x86_64-disk.img $PWD/test1234
$ sudo rm test1234 # without force
rm: cannot remove 'test1234': Device or resource busy
$ sudo rm test1234 -f # with force
rm: cannot remove 'test1234': Device or resource busy
```

How does the kubelet manage to delete the file?

Reply:
Right you can replicate the issue pretty easily doing the following, in main comment out these lines: https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/hotplug-disk/mount.go#L719-L726 which is the code that unmounts the bindmounted volumes normally, this is to force the kubelet to win the 'race'. Then create a VM, create a volume (preferably a local volume, so you see the file disappearing in the local volume directory), then hotplug that volume into the running VM. So far this will be normal. Now locate the virt-launcher pod and force delete the pod (--force --grace-period=0). Because we commented out the code that unmounts the volume, they will still exist in the emptyDir. Then the kubelet will run and wipe out the disk.img. You can verify that the disk.img is gone in the local volume.

Reply:
I am not sure if that is a sufficient explanation. Here an idea what I would find great if you could verify: My suspicion is more that the disk.img still exists (and that qemu is still connected to it), but that the kubelet simply unmounted its `emptyDir` bindmount.

If so, it would be sufficient so not try to unmount the `disk.img` from this directory, but instead, when doing the initial `disk.img` mount, recording the actual `emptyDir` location and  doing the unmount there. I would not be surprised (but of course has to be proven), if you would see errors in the `kubelet` logs where it has issues actually deleting the `emptyDir` in the temporary data location.

Reply:
So this is why I said use a local volume. Because you can then ssh into the node, and switch to the directory that backs the SOURCE volume (the one attached to the attachment pod, which is then bind mounted into the emptyDir). When the emptyDir is cleared, it also removes the disk.img from the SOURCE path, aka the original disk.img in the original volume. So to be 100% clear, the disk.img that exists in the volume that it hotplugged as attached to the attachment volume is removed when the emptyDir is cleared.

This was first reported here: https://groups.google.com/d/msgid/kubevirt-dev/c6e5b2ef-d6a1-4546-966e-6afae641e2a0n%40googlegroups.com?utm_medium=email&utm_source=footer I was unable to reproduce the race myself, but if I comment out the unmount code in hotplug I can reproduce without fail. In the report they say they are using a modified kubelet where in the code that clears the emptyDir they added some code to unmount any mounted volumes, and that solved the problem for them.

I just verified that I am bind mounting the directory, not the disk.img file. I could bind the file itself and see if that triggers the resource busy. That is definitely an experiment we can do.

Reply:
>When the emptyDir is cleared, it also removes the disk.img from the SOURCE path, aka the original disk.img in the original volume.

Again, as demonstrated above, I can't delete a `qemu.img` if it is bind-mounted. It is not clear to me how the kubelet could achieve that.

Reply:
@awels in the hotplug code, do you bind-mount the file (like we do in the containerDisk case) or do you bind-mount a directory? If you bind-mount the directory and not the file itself that may be the issue in the hotplug code.

Reply:
> I just verified that I am bind mounting the directory, not the disk.img file. I could bind the file itself and see if that triggers the resource busy. That is definitely an experiment we can do.

Its the directory not the file.

Reply:
I am going to do that experiment some time today I think.

Reply:
> Its the directory not the file.

I thin t his is your main issue.

Reply:
https://github.com/kubevirt/kubevirt/pull/6464 implements mounting the file instead of the directory, and that does exactly what you said it would.